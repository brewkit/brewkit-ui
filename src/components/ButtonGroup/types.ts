import { ButtonGroupProps } from '@material-ui/core/ButtonGroup';


export type Props = ButtonGroupProps;
