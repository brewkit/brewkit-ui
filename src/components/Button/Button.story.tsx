import * as React from 'react';
import { Story } from '@storybook/react';
import ArrowBack from '@material-ui/icons/ArrowBack';
import AddCircleOutline from '@material-ui/icons/AddCircleOutline';
import { ButtonProps } from './Button.types';

/**
 * This is an annoying hack to get Storybook to properly parse the documentation when the underlying component
 * is wrapped in a HOC. Expect to do/see this on every component with MUI theming options.
 */
import { Button } from './Button';

export default {
    component: Button,
    title: 'Inputs/Button',
};

const defaultArgs = {
    disabled: false,
    disableElevation: true,
    fullWidth: false,
    children: 'I am button text',
};

export const Sandbox: Story<ButtonProps> = function sandbox(args) {
    return <Button {...args} />;
};

Sandbox.args = {
    ...defaultArgs,
    color: 'primary',
};

export const General: Story<ButtonProps> = function storyGeneral(args) {
    return <Button {...args} />;
};

General.args = {
    ...defaultArgs,
    color: 'primary',
    variant: 'text',
};

export const WithStartIcon: Story<ButtonProps> = function storyWithStartIcon(
    args,
) {
    return <Button {...args} />;
};

WithStartIcon.args = {
    ...defaultArgs,
    color: 'primary',
    startIcon: <ArrowBack />,
};

export const WithEndIcon: Story<ButtonProps> = function storyWithEndIcon(args) {
    return <Button {...args} />;
};

WithEndIcon.args = {
    ...defaultArgs,
    color: 'primary',
    endIcon: <AddCircleOutline />,
};

export const OnlyIcon: Story<ButtonProps> = function storyOnlyIcon(args) {
    return <Button {...args} />;
};

OnlyIcon.args = {
    ...defaultArgs,
    color: 'primary',
    children: <AddCircleOutline />,
};

export const Disabled: Story<ButtonProps> = function storyDisabled(args) {
    return <Button {...args} />;
};

Disabled.args = {
    ...defaultArgs,
    color: 'primary',
    disabled: true,
};

export const Loading: Story<ButtonProps> = function storyLoading(args) {
    return <Button {...args} />;
};

Loading.args = {
    ...defaultArgs,
    color: 'primary',
    disabled: true,
    loading: true,
};
