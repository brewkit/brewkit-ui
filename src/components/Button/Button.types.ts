import * as React from 'react';
import { ButtonProps as MUIButtonProps } from '@material-ui/core';


/**
 * Declare any theme variables we want to make available.
 */
declare module '@material-ui/core/styles/createTheme' {
    interface Theme {
        BkButton: {
            loading: React.CSSProperties,
        },
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        BkButton?: {
            loading?: React.CSSProperties,
        },
    }
}


export type ButtonProps = MUIButtonProps & {

    /**
     * If `true`, the button will be disabled and show a spinner.
     * @default false
     */
    loading?: boolean,


    /**
     * @ignore
     */
    classes?: Record<string, any>,

};
