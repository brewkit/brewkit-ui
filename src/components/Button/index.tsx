import * as React from 'react';
import clsx from 'clsx';
import { withStyles } from '@material-ui/core/styles';
import MuiButton from '@material-ui/core/Button';
import CircularProgress from '@components/CircularProgress';
import styles from './styles';
import { Props } from './types';


/**
 * Buttons allow users to take actions and make choices with a single tap.
 *
 * [Material-UI Docs](https://material-ui.com/components/buttons/)
 *
 *
 * ## Differences in Brewkit:
 *
 * - Adds a `loading` prop that automatically disables the button and shows a loading state. Loading class can be
 * styled with `BkButton.loading` within your theme.
 */
export const Button = React.forwardRef(({
    loading,
    disabled,
    startIcon,
    endIcon,
    children,
    className,
    classes,
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => {


    /**
     * merge our custom classes? into the ones passed to the component
     */
    const buttonClasses = clsx(
        classes?.root,
        loading && classes?.loading,
        className,
    );


    return (
        <MuiButton
            {...otherProps}
            className={buttonClasses}
            disabled={disabled ?? loading}
            ref={ref}
        >

            <div className={classes?.content}>
                {startIcon}
                {children && <span>{children}</span>}
                {endIcon}
            </div>

            <div className={classes?.loader}>
                <CircularProgress color="primary" size="1rem" thickness={5} />
            </div>

        </MuiButton>
    );


});


Button.displayName = 'Button';


export default withStyles(styles)(Button);
