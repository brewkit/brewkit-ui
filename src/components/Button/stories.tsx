import * as React from 'react';
import { withStyles } from '@material-ui/core/styles';
import { text, select, boolean } from '@storybook/addon-knobs';
import ArrowBack from '@material-ui/icons/ArrowBack';
import AddCircleOutline from '@material-ui/icons/AddCircleOutline';


/**
 * This is an annoying hack to get Storybook to properly parse the documentation when the underlying component
 * is wrapped in a HOC. Expect to do/see this on every component with MUI theming options.
 */
import { Button as Component } from './index';
import styles from './styles';
const Button = withStyles(styles)(Component)


export default {
    component: Component,
    title: 'Inputs/Button',
};


export const Sandbox = (): React.ReactElement => {

    const color = select('color', ['default', 'primary', 'secondary', 'inherit'], 'primary');
    const disabled = boolean('disabled', false);
    const disableElevation = boolean('disableElevation', true);
    const fullWidth = boolean('fullWidth', false);
    const loading = boolean('loading', false);
    const variant = select('variant', ['contained', 'outlined', 'text'], 'contained');
    const size = select('size', ['large', 'medium', 'small'], 'medium');
    const children = text('children', 'default');

    return (
        <Button
            color={color}
            disabled={disabled}
            disableElevation={disableElevation}
            fullWidth={fullWidth}
            loading={loading}
            variant={variant}
            size={size}
            children={children}
        />
    );

};


export const General = (): React.ReactElement => (
    <div style={{ display: 'grid', gridAutoFlow: 'column', gridGap: '1rem' }}>
        <Button>Generic Button</Button>
        <Button variant="text">Generic Button</Button>
    </div>
)


export const WithStartIcon = (): React.ReactElement => (
    <Button startIcon={<ArrowBack />}>Generic Button</Button>
)


export const WithEndIcon = (): React.ReactElement => (
    <Button endIcon={<AddCircleOutline />}>Generic Button</Button>
)


export const OnlyIcon = (): React.ReactElement => (
    <Button>
        <AddCircleOutline />
    </Button>
)


export const Disabled = (): React.ReactElement => (
    <div style={{ display: 'grid', gridAutoFlow: 'column', gridGap: '1rem' }}>
        <Button disabled>Generic Button</Button>
        <Button variant="text" disabled>Generic Button</Button>
    </div>
)


export const Loading = (): React.ReactElement => (
    <div style={{ display: 'grid', gridAutoFlow: 'column', gridGap: '1rem' }}>
        <Button loading>Generic Button</Button>
        <Button variant="text" loading>Generic Button</Button>
    </div>
)
