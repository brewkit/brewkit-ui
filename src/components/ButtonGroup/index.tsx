import * as React from 'react';
import MuiButtonGroup from '@material-ui/core/ButtonGroup';
import { Props } from './types';


/**
 * The ButtonGroup component can be used to group related buttons.
 *
 * [Material-UI Docs](https://material-ui.com/components/button-group/)
 *
 * > No differences in Brewkit.
 */
export const ButtonGroup = React.forwardRef(({
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => (


    <MuiButtonGroup ref={ref} {...otherProps} />


));


ButtonGroup.displayName = 'ButtonGroup';


export default ButtonGroup;
