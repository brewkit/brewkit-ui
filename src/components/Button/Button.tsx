import * as React from 'react';
import clsx from 'clsx';

import MuiButton from '@material-ui/core/Button';

import { CircularProgress } from '../CircularProgress';

import useStyles from './Button.styles';
import { ButtonProps } from './Button.types';

/**
 * Buttons allow users to take actions and make choices with a single tap.
 *
 * [Material-UI Docs](https://material-ui.com/components/buttons/)
 *
 *
 * ## Differences in Brewkit:
 *
 * - Adds a `loading` prop that automatically disables the button and shows a loading state. Loading class can be
 * styled with `BkButton.loading` within your theme.
 */
export const Button = React.forwardRef(
    (
        props: ButtonProps,
        ref: React.Ref<HTMLButtonElement>,
    ): React.ReactElement => {
        const {
            loading,
            disabled,
            startIcon,
            endIcon,
            children,
            className,
            ...otherProps
        } = props;

        const classes = useStyles();

        /**
         * merge our custom classes? into the ones passed to the component
         */
        const buttonClasses = clsx(
            classes?.root,
            loading && classes?.loading,
            className,
        );

        return (
            <MuiButton
                {...otherProps}
                className={buttonClasses}
                disabled={disabled ?? loading}
                ref={ref}>
                <div className={classes?.content}>
                    {startIcon}
                    {children && <span>{children}</span>}
                    {endIcon}
                </div>

                <div className={classes?.loader}>
                    <CircularProgress
                        color="primary"
                        size="1rem"
                        thickness={5}
                    />
                </div>
            </MuiButton>
        );
    },
);

Button.displayName = 'Button';

export default Button;
