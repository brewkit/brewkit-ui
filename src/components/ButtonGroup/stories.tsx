import * as React from 'react';
import {text, select, boolean} from '@storybook/addon-knobs';
import ButtonGroup from '@components/ButtonGroup';
import Button from '@components/Button';


export default {
    component: ButtonGroup,
    parameters: {
        // componentSubtitle: <Flag color="success">Stable</Flag>,
    },
    title: 'Inputs/ButtonGroup',
};


export const Sandbox = (): React.ReactElement => {

    const color = select('color', ['default', 'primary', 'secondary', 'inherit'], 'primary');
    const disabled = boolean('disabled', false);
    const disableElevation = boolean('disableElevation', true);
    const variant = select('variant', ['contained', 'outlined', 'text'], 'contained');
    const orientation = select('orientation', ['vertical', 'horizontal'], 'horizontal');

    return (
        <ButtonGroup
            variant={variant}
            color={color}
            disabled={disabled}
            disableElevation={disableElevation}
            orientation={orientation}
        >
            <Button>One</Button>
            <Button>Two</Button>
            <Button>Three</Button>
        </ButtonGroup>
    );

};
