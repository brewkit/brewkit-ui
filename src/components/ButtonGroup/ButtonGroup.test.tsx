// @ts-nocheck
import * as React from 'react';
import { create } from 'react-test-renderer';
import ButtonGroup from '.';
import Button from '../Button';

it('Renders ButtonGroup component', () => {
    const comp = create(
        <ButtonGroup>
            <Button>One</Button>
            <Button>Two</Button>
            <Button>Three</Button>
        </ButtonGroup>
    ).toJSON();

    expect(comp).toMatchSnapshot();
});
