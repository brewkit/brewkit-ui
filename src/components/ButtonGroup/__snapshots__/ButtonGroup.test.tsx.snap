// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders ButtonGroup component 1`] = `
<div
  className="MuiButtonGroup-root"
  role="group"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-outlined Button-root-1 MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      <div
        className="Button-content-2"
      >
        <span>
          One
        </span>
      </div>
      <div
        className="Button-loader-3"
      >
        <div
          className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style={
            Object {
              "height": "1rem",
              "width": "1rem",
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx={44}
              cy={44}
              fill="none"
              r={19.5}
              strokeWidth={5}
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </span>
  </button>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-outlined Button-root-1 MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      <div
        className="Button-content-2"
      >
        <span>
          Two
        </span>
      </div>
      <div
        className="Button-loader-3"
      >
        <div
          className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style={
            Object {
              "height": "1rem",
              "width": "1rem",
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx={44}
              cy={44}
              fill="none"
              r={19.5}
              strokeWidth={5}
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </span>
  </button>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-outlined Button-root-1 MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      <div
        className="Button-content-2"
      >
        <span>
          Three
        </span>
      </div>
      <div
        className="Button-loader-3"
      >
        <div
          className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style={
            Object {
              "height": "1rem",
              "width": "1rem",
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx={44}
              cy={44}
              fill="none"
              r={19.5}
              strokeWidth={5}
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </span>
  </button>
</div>
`;
